### ISSUE TITLE
[imgui] issue with opengl bindings

### ISSUE THREAD

**Issue by cloudhan** [microsoft/vcpkg#14753]
****

**Is your feature request related to a problem? Please describe.**
`vcpkg install imgui[opengl3-glbinding-binding,glfw-binding]:x64-windows` will build with glew.

resulting
```
[build] imguid.lib(imgui_impl_opengl3.cpp.obj) : error LNK2019: unresolved external symbol __imp___glewActiveTexture referenced in function "void __cdecl ImGui_ImplOpenGL3_RenderDrawData(struct ImDrawData *)" (?ImGui_ImplOpenGL3_RenderDrawData@@YAXPEAUImDrawData@@@Z) [****.vcxproj]
[build] imguid.lib(imgui_impl_opengl3.cpp.obj) : error LNK2019: unresolved external symbol __imp___glewBlendEquation referenced in function "void __cdecl ImGui_ImplOpenGL3_SetupRenderState(struct ImDrawData *,int,int,unsigned int)" (?ImGui_ImplOpenGL3_SetupRenderState@@YAXPEAUImDrawData@@HHI@Z) [****.vcxproj]
[build] imguid.lib(imgui_impl_opengl3.cpp.obj) : error LNK2019: unresolved external symbol __imp___glewBlendFuncSeparate referenced in function "void __cdecl ImGui_ImplOpenGL3_RenderDrawData(struct ImDrawData *)" (?ImGui_ImplOpenGL3_RenderDrawData@@YAXPEAUImDrawData@@@Z) [****.vcxproj]
[build] imguid.lib(imgui_impl_opengl3.cpp.obj) : error LNK2019: unresolved external symbol __imp___glewBindBuffer referenced in function "void __cdecl ImGui_ImplOpenGL3_RenderDrawData(struct ImDrawData *)" (?ImGui_ImplOpenGL3_RenderDrawData@@YAXPEAUImDrawData@@@Z) [****.vcxproj]
[build] imguid.lib(imgui_impl_opengl3.cpp.obj) : error LNK2019: unresolved external symbol __imp___glewBufferData referenced in function "void __cdecl ImGui_ImplOpenGL3_RenderDrawData(struct ImDrawData *)" (?ImGui_ImplOpenGL3_RenderDrawData@@YAXPEAUImDrawData@@@Z) [****.vcxproj]
[build] imguid.lib(imgui_impl_opengl3.cpp.obj) : error LNK2019: unresolved external symbol __imp___glewDeleteBuffers referenced in function "void __cdecl ImGui_ImplOpenGL3_DestroyDeviceObjects(void)" (?ImGui_ImplOpenGL3_DestroyDeviceObjects@@YAXXZ) [****.vcxproj]
[build] imguid.lib(imgui_impl_opengl3.cpp.obj) : error LNK2019: unresolved external symbol __imp___glewGenBuffers referenced in function "bool __cdecl ImGui_ImplOpenGL3_CreateDeviceObjects(void)" (?ImGui_ImplOpenGL3_CreateDeviceObjects@@YA_NXZ) [****.vcxproj]
[build] imguid.lib(imgui_impl_opengl3.cpp.obj) : error LNK2019: unresolved external symbol __imp___glewAttachShader referenced in function "bool __cdecl ImGui_ImplOpenGL3_CreateDeviceObjects(void)" (?ImGui_ImplOpenGL3_CreateDeviceObjects@@YA_NXZ) [****.vcxproj]
[build] imguid.lib(imgui_impl_opengl3.cpp.obj) : error LNK2019: unresolved external symbol __imp___glewBlendEquationSeparate referenced in function "void __cdecl ImGui_ImplOpenGL3_RenderDrawData(struct ImDrawData *)" (?ImGui_ImplOpenGL3_RenderDrawData@@YAXPEAUImDrawData@@@Z) [****.vcxproj]
[build] imguid.lib(imgui_impl_opengl3.cpp.obj) : error LNK2019: unresolved external symbol __imp___glewCompileShader referenced in function "bool __cdecl ImGui_ImplOpenGL3_CreateDeviceObjects(void)" (?ImGui_ImplOpenGL3_CreateDeviceObjects@@YA_NXZ) [****.vcxproj]
[build] imguid.lib(imgui_impl_opengl3.cpp.obj) : error LNK2019: unresolved external symbol __imp___glewCreateProgram referenced in function "bool __cdecl ImGui_ImplOpenGL3_CreateDeviceObjects(void)" (?ImGui_ImplOpenGL3_CreateDeviceObjects@@YA_NXZ) [****.vcxproj]
[build] imguid.lib(imgui_impl_opengl3.cpp.obj) : error LNK2019: unresolved external symbol __imp___glewCreateShader referenced in function "bool __cdecl ImGui_ImplOpenGL3_CreateDeviceObjects(void)" (?ImGui_ImplOpenGL3_CreateDeviceObjects@@YA_NXZ) [****.vcxproj]
[build] imguid.lib(imgui_impl_opengl3.cpp.obj) : error LNK2019: unresolved external symbol __imp___glewDeleteProgram referenced in function "void __cdecl ImGui_ImplOpenGL3_DestroyDeviceObjects(void)" (?ImGui_ImplOpenGL3_DestroyDeviceObjects@@YAXXZ) [****.vcxproj]
[build] imguid.lib(imgui_impl_opengl3.cpp.obj) : error LNK2019: unresolved external symbol __imp___glewDeleteShader referenced in function "void __cdecl ImGui_ImplOpenGL3_DestroyDeviceObjects(void)" (?ImGui_ImplOpenGL3_DestroyDeviceObjects@@YAXXZ) [****.vcxproj]
[build] imguid.lib(imgui_impl_opengl3.cpp.obj) : error LNK2019: unresolved external symbol __imp___glewDetachShader referenced in function "void __cdecl ImGui_ImplOpenGL3_DestroyDeviceObjects(void)" (?ImGui_ImplOpenGL3_DestroyDeviceObjects@@YAXXZ) [****.vcxproj]
[build] imguid.lib(imgui_impl_opengl3.cpp.obj) : error LNK2019: unresolved external symbol __imp___glewEnableVertexAttribArray referenced in function "void __cdecl ImGui_ImplOpenGL3_SetupRenderState(struct ImDrawData *,int,int,unsigned int)" (?ImGui_ImplOpenGL3_SetupRenderState@@YAXPEAUImDrawData@@HHI@Z) [****.vcxproj]
[build] imguid.lib(imgui_impl_opengl3.cpp.obj) : error LNK2019: unresolved external symbol __imp___glewGetAttribLocation referenced in function "bool __cdecl ImGui_ImplOpenGL3_CreateDeviceObjects(void)" (?ImGui_ImplOpenGL3_CreateDeviceObjects@@YA_NXZ) [****.vcxproj]
[build] imguid.lib(imgui_impl_opengl3.cpp.obj) : error LNK2019: unresolved external symbol __imp___glewGetProgramInfoLog referenced in function "bool __cdecl CheckProgram(unsigned int,char const *)" (?CheckProgram@@YA_NIPEBD@Z) [****.vcxproj]
[build] imguid.lib(imgui_impl_opengl3.cpp.obj) : error LNK2019: unresolved external symbol __imp___glewGetProgramiv referenced in function "bool __cdecl CheckProgram(unsigned int,char const *)" (?CheckProgram@@YA_NIPEBD@Z) [****.vcxproj]
[build] imguid.lib(imgui_impl_opengl3.cpp.obj) : error LNK2019: unresolved external symbol __imp___glewGetShaderInfoLog referenced in function "bool __cdecl CheckShader(unsigned int,char const *)" (?CheckShader@@YA_NIPEBD@Z) [****.vcxproj]
[build] imguid.lib(imgui_impl_opengl3.cpp.obj) : error LNK2019: unresolved external symbol __imp___glewGetShaderiv referenced in function "bool __cdecl CheckShader(unsigned int,char const *)" (?CheckShader@@YA_NIPEBD@Z) [****.vcxproj]
[build] imguid.lib(imgui_impl_opengl3.cpp.obj) : error LNK2019: unresolved external symbol __imp___glewGetUniformLocation referenced in function "bool __cdecl ImGui_ImplOpenGL3_CreateDeviceObjects(void)" (?ImGui_ImplOpenGL3_CreateDeviceObjects@@YA_NXZ) [****.vcxproj]
[build] imguid.lib(imgui_impl_opengl3.cpp.obj) : error LNK2019: unresolved external symbol __imp___glewLinkProgram referenced in function "bool __cdecl ImGui_ImplOpenGL3_CreateDeviceObjects(void)" (?ImGui_ImplOpenGL3_CreateDeviceObjects@@YA_NXZ) [****.vcxproj]
[build] imguid.lib(imgui_impl_opengl3.cpp.obj) : error LNK2019: unresolved external symbol __imp___glewShaderSource referenced in function "bool __cdecl ImGui_ImplOpenGL3_CreateDeviceObjects(void)" (?ImGui_ImplOpenGL3_CreateDeviceObjects@@YA_NXZ) [****.vcxproj]
[build] imguid.lib(imgui_impl_opengl3.cpp.obj) : error LNK2019: unresolved external symbol __imp___glewUniform1i referenced in function "void __cdecl ImGui_ImplOpenGL3_SetupRenderState(struct ImDrawData *,int,int,unsigned int)" (?ImGui_ImplOpenGL3_SetupRenderState@@YAXPEAUImDrawData@@HHI@Z) [****.vcxproj]
[build] imguid.lib(imgui_impl_opengl3.cpp.obj) : error LNK2019: unresolved external symbol __imp___glewUniformMatrix4fv referenced in function "void __cdecl ImGui_ImplOpenGL3_SetupRenderState(struct ImDrawData *,int,int,unsigned int)" (?ImGui_ImplOpenGL3_SetupRenderState@@YAXPEAUImDrawData@@HHI@Z) [****.vcxproj]
[build] imguid.lib(imgui_impl_opengl3.cpp.obj) : error LNK2019: unresolved external symbol __imp___glewUseProgram referenced in function "void __cdecl ImGui_ImplOpenGL3_RenderDrawData(struct ImDrawData *)" (?ImGui_ImplOpenGL3_RenderDrawData@@YAXPEAUImDrawData@@@Z) [****.vcxproj]
[build] imguid.lib(imgui_impl_opengl3.cpp.obj) : error LNK2019: unresolved external symbol __imp___glewVertexAttribPointer referenced in function "void __cdecl ImGui_ImplOpenGL3_SetupRenderState(struct ImDrawData *,int,int,unsigned int)" (?ImGui_ImplOpenGL3_SetupRenderState@@YAXPEAUImDrawData@@HHI@Z) [****.vcxproj]
[build] imguid.lib(imgui_impl_opengl3.cpp.obj) : error LNK2019: unresolved external symbol __imp___glewDrawElementsBaseVertex referenced in function "void __cdecl ImGui_ImplOpenGL3_RenderDrawData(struct ImDrawData *)" (?ImGui_ImplOpenGL3_RenderDrawData@@YAXPEAUImDrawData@@@Z) [****.vcxproj]
[build] imguid.lib(imgui_impl_opengl3.cpp.obj) : error LNK2019: unresolved external symbol __imp___glewBindSampler referenced in function "void __cdecl ImGui_ImplOpenGL3_RenderDrawData(struct ImDrawData *)" (?ImGui_ImplOpenGL3_RenderDrawData@@YAXPEAUImDrawData@@@Z) [****.vcxproj]
[build] imguid.lib(imgui_impl_opengl3.cpp.obj) : error LNK2019: unresolved external symbol __imp___glewBindVertexArray referenced in function "void __cdecl ImGui_ImplOpenGL3_RenderDrawData(struct ImDrawData *)" (?ImGui_ImplOpenGL3_RenderDrawData@@YAXPEAUImDrawData@@@Z) [****.vcxproj]
[build] imguid.lib(imgui_impl_opengl3.cpp.obj) : error LNK2019: unresolved external symbol __imp___glewDeleteVertexArrays referenced in function "void __cdecl ImGui_ImplOpenGL3_RenderDrawData(struct ImDrawData *)" (?ImGui_ImplOpenGL3_RenderDrawData@@YAXPEAUImDrawData@@@Z) [****.vcxproj]
[build] imguid.lib(imgui_impl_opengl3.cpp.obj) : error LNK2019: unresolved external symbol __imp___glewGenVertexArrays referenced in function "void __cdecl ImGui_ImplOpenGL3_RenderDrawData(struct ImDrawData *)" (?ImGui_ImplOpenGL3_RenderDrawData@@YAXPEAUImDrawData@@@Z) [****.vcxproj]
```

**Proposed solution**

So lets make the problem clear first. For opengl there are two types of helper library commonly used.

1.  OpenGL symbol loader: e.g. glbinding, glew, gl3w, etc.
2. OpenGL window and context helper: e.g. glfw, glut

In this case, imgui[opengl3-glbinding-binding,glfw-binding] is actually built as imgui[opengl3-glew-binding,glfw-binding], resulting glew symbols in the `imguid.lib` file.

This should be fix.

**Describe alternatives you've considered**

Even the above issue is fixed, there will be potential problem:
1. project A use glew +  glfw
2. project B use glbinding + glfw
3. project C use DX12
4. ...

since there will be no way to let one installation of vcpkg to supply polymorphic `imgui.lib`, this problem will never be solved if vcpkg building bindings into library code.

The only solution I can come up with is copying bindings `imgui_impl_*.cpp` files into imgui's `include` folder. Let the use do

```
// imgui_impl.cpp file
// assuming the user use glbinding + glfw

// user do necessary definition

#include <bindings/imgui_impl_glfw.cpp>
#include <bindings/imgui_impl_opengl3.cpp>
```
and let the user compile this source and link the obj file to its executable.

**Comment by JackBoosY** [comment]
Is this a build issue or usage issue? I can build imgui succesfully.
Imgui use unofficial CMakeLists.txt to be built currentlly, and it has some problems. See https://github.com/ocornut/imgui/pull/1713

**Comment by cloudhan** [comment]
Build Issue. glew symbol is included, when it should not.

**Comment by cloudhan** [comment]
See this minimum reproduce.

[issue-14753.zip](https://github.com/microsoft/vcpkg/files/5596912/issue-14753.zip)

**Comment by JackBoosY** [comment]
Confirmed.

**Comment by JackBoosY** [comment]
@RT222 Seems most of these binding features are missing export the dependency libraries. Can you confirm this?

**Comment by RT222** [comment]
I'm sorry about this, but this problem is beyond my comprehension, and I currently lack the time to dig into it. If it helps, I can revert my changes to get back to the old fashioned way of including the sources files manually.

**Comment by JackBoosY** [comment]
@RT222  Fine, I will handle this.

**Comment by xarthurx** [comment]
Same issue here.
Using `libigl` with `imgui` support, and got exactly the same erroes.

**Comment by MrSimbax** [comment]
### Problem

I have both GLEW and Glad installed by Vcpkg, as I need them both for different projects. Dear ImGui doesn't work for the project with Glad (and it was working fine before).

The latter project uses the Glad loader with SDL2 window therefore I used the following command to install Dear ImGui:

```
./vcpkg.exe install imgui[sdl2-binding,opengl3-glad-binding]
```

However, this resulted in linker errors similar to the ones in the OP's post. `imguid.lib` contained some Glew symbols. So I tried installing the Glew bindings too even though my application doesn't use Glew.

```
./vcpkg.exe install imgui[sdl2-binding,opengl3-glad-binding,opengl3-glew-binding]
```

My application then linked fine but it crashed on [`ImGui_ImplOpenGL3_NewFrame()`](https://github.com/ocornut/imgui/blob/6a161b878943241ddecbeee4ae27103dbd2d33d0/backends/imgui_impl_opengl3.cpp#L240) with access violation exception.

### Root cause

I've found out with Visual Studio debugger that `ImGui_ImplOpenGL3_NewFrame()` was trying to call [`glCreateShader()`](https://github.com/ocornut/imgui/blob/6a161b878943241ddecbeee4ae27103dbd2d33d0/backends/imgui_impl_opengl3.cpp#L670) but it expanded to wrong memory address 0 (`nullptr`). This usually means that the loader library wasn't properly initialized with a function like `gladLoadGLLoader()` beforehand but I found no such issue in my code.

Turns out that ImGui was compiled by Vcpkg with wrong header included in [imgui_impl_opengl3.cpp](https://github.com/ocornut/imgui/blob/6a161b878943241ddecbeee4ae27103dbd2d33d0/backends/imgui_impl_opengl3.cpp#L99-L128).

The question is, why was the GLEW header used instead of Glad header? So it seems that the OpenGL loader is [automatically detected](https://github.com/ocornut/imgui/blob/b1a18d82e32f13a2ae62df70d08ee46bc8ee6a76/backends/imgui_impl_opengl3.h#L68-L71) by preprocessor based on available header files. Unfortunately, due to the order of `if`s it seems Vcpkg compiled the ImGui with the GLEW bindings instead of the Glad bindings.

### Workaround

I guess uninstalling Glew or using separate Vcpkg installation with only one loader installed would work. However, I find those workarounds too annoying and managed to find another way to fix it. See the following change in Vcpkg:

```diff
diff --git a/ports/imgui/CMakeLists.txt b/ports/imgui/CMakeLists.txt
index f6f85b29e..4aa77096e 100644
--- a/ports/imgui/CMakeLists.txt
+++ b/ports/imgui/CMakeLists.txt
@@ -81,6 +81,7 @@ endif()
 if(IMGUI_BUILD_OPENGL3_GLAD_BINDING)
     find_package(glad CONFIG REQUIRED)
     target_link_libraries(${PROJECT_NAME} PUBLIC glad::glad)
+    target_compile_definitions(${PROJECT_NAME} PUBLIC IMGUI_IMPL_OPENGL_LOADER_GLAD)
     target_sources(${PROJECT_NAME} PRIVATE ${CMAKE_CURRENT_SOURCE_DIR}/backends/imgui_impl_opengl3.cpp)
 endif()
```

I hope it helps somebody with a similar issue.

Also, please consider adding the compile definitions to all loaders. It will ensure that the correct variant is installed even when there are multiple OpenGL loaders installed. Another thing to consider is to somehow make it so that multiple ImGui variants can be installed at the same time in one Vcpkg installation but that's another issue.

**Comment by JackBoosY** [comment]
@MrSimbax According to the code in _exmaples/example_glfw_opengl3/main/cpp_ line 83-99:
```cpp
#if defined(IMGUI_IMPL_OPENGL_LOADER_GL3W)
    bool err = gl3wInit() != 0;
#elif defined(IMGUI_IMPL_OPENGL_LOADER_GLEW)
    bool err = glewInit() != GLEW_OK;
#elif defined(IMGUI_IMPL_OPENGL_LOADER_GLAD)
    bool err = gladLoadGL() == 0;
#elif defined(IMGUI_IMPL_OPENGL_LOADER_GLAD2)
    bool err = gladLoadGL(glfwGetProcAddress) == 0; // glad2 recommend using the windowing library loader instead of the (optionally) bundled one.
#elif defined(IMGUI_IMPL_OPENGL_LOADER_GLBINDING2)
    bool err = false;
    glbinding::Binding::initialize();
#elif defined(IMGUI_IMPL_OPENGL_LOADER_GLBINDING3)
    bool err = false;
    glbinding::initialize([](const char* name) { return (glbinding::ProcAddress)glfwGetProcAddress(name); });
#else
    bool err = false; // If you use IMGUI_IMPL_OPENGL_LOADER_CUSTOM, your loader is likely to requires some form of initialization.
#endif
```
Does this confirm that these features are in conflict?

**Comment by mathisloge** [comment]
@JackBoosY https://github.com/ocornut/imgui/blob/ad5d1a8429ea219d3d34e6a36a48918650402697/backends/imgui_impl_opengl3.h#L68 this would led to the definition of the flags.

**Comment by JackBoosY** [comment]
@mathisloge Therefore, these definitions should not be added to the compilation options, because they will be automatically defined.

**Comment by mathisloge** [comment]
@JackBoosY  i think @Dylan-Davies has a point there (https://github.com/microsoft/vcpkg/issues/19231#issuecomment-889712512). Since these preprocessors would be evaluated at vcpkg-build time. 
So we could at opengl targets (since it only affects opengl) which all specify the correct compile definitions. 
e.g. 
```
add_library(imgui-glew ${CMAKE_CURRENT_SOURCE_DIR}/backends/imgui_impl_opengl3.cpp)
target_compile_definitions(imgui-glew PUBLIC IMGUI_IMPL_OPENGL_LOADER_GLEW)

add_library(imgui-glad ${CMAKE_CURRENT_SOURCE_DIR}/backends/imgui_impl_opengl3.cpp)
target_compile_definitions(imgui-glew PUBLIC IMGUI_IMPL_OPENGL_LOADER_GLAD)

and so on.
```

**Comment by mathisloge** [comment]
Just as a sidenode: if there is a loader defined beforehand, the detection would be disabled.
https://github.com/ocornut/imgui/blob/ad5d1a8429ea219d3d34e6a36a48918650402697/backends/imgui_impl_opengl3.h#L47

**Comment by JackBoosY** [comment]
@mathisloge Make sence, but we should wait for the upstream reply first.

**Comment by mathisloge** [comment]
anyone pinged upstream? 😄

**Comment by JackBoosY** [comment]
@dg0yt what do you think about since you are the contributor.

**Comment by dg0yt** [comment]
> @dg0yt what do you think about since you are the contributor.

@JackBoosY Contributor of what?

**Comment by JackBoosY** [comment]
> > @dg0yt what do you think about since you are the contributor.
> 
> @JackBoosY Contributor of what?

Sorry for ping the wrong guy.

**Comment by JackBoosY** [comment]
cc @ocornut, the imgui owner.

**Comment by ocornut** [comment]
Hello, I posted an answer at https://github.com/ocornut/imgui/issues/4390

**Comment by mathisloge** [comment]
@JackBoosY so with the answer ocornut gave: 
Should I open a PR that adds new targets for each of the opengl loaders? 
Or do you already working on it?

**Comment by JackBoosY** [comment]
@mathisloge  Will fix later.

**Comment by JackBoosY** [comment]
@mathisloge  Yes, I will fix this in the nearly feature.

**Comment by ocornut** [comment]
We decided to bite the bullet and we are working toward embedding our own minimalist OpenGL loader into the backend, in order to put in end to that endless amount of confusion/misery caused by the OpenGL loaders.

We already have a version working and we'll either aim to use it in 1.84 or right after 1.84 is released.

TL;DR; ihmo don't put effort into solving this on your end.

**Comment by ocornut** [comment]
The dear imgui opengl backend now embeds its own loader, simplifying a lot the trouble people had and hte packaging problem for vcpkg. See https://github.com/ocornut/imgui/issues/4445

**Comment by JackBoosY** [comment]
Close this PR until upstream fixes this.
